server/routerlicious/packages/services-ordering-rdkafka/src/rdkafkaConsumer.ts

```diff
@@ -7,484 +7,489 @@ import type * as kafkaTypes from "node-rdkafka";
-		}
-
-		const partitionsWithEpoch: IPartitionWithEpoch[] = [];
-
-		for (let i = 0; i < partitions.length; ++i) {
-			const partitionWithEpoch = partitions[i] as IPartitionWithEpoch;
-			partitionWithEpoch.leaderEpoch = epochs[i];
-			partitionsWithEpoch.push(partitionWithEpoch);
-		}
-
-		return partitionsWithEpoch;
-	}
+    private readonly consumerOptions: IKafkaConsumerOptions;
+    private consumer?: kafkaTypes.KafkaConsumer;
+    private zooKeeperClient?: IZookeeperClient;
+    private closed = false;
+    private isRebalancing = true;
+    private assignedPartitions: Set<number> = new Set();
+    private readonly pendingCommits: Map<number, Deferred<void>> = new Map();
+    private readonly pendingMessages: Map<number, kafkaTypes.Message[]> = new Map();
+    private readonly latestOffsets: Map<number, number> = new Map();
+
+    constructor(
+        endpoints: IKafkaEndpoints,
+        clientId: string,
+        topic: string,
+        public readonly groupId: string,
+        options?: Partial<IKafkaConsumerOptions>) {
+        super(endpoints, clientId, topic, options);
+
+        this.defaultRestartOnKafkaErrorCodes = [
+            this.kafka.CODES.ERRORS.ERR__TRANSPORT,
+            this.kafka.CODES.ERRORS.ERR__MSG_TIMED_OUT,
+            this.kafka.CODES.ERRORS.ERR__ALL_BROKERS_DOWN,
+            this.kafka.CODES.ERRORS.ERR__TIMED_OUT,
+            this.kafka.CODES.ERRORS.ERR__SSL,
+            this.kafka.CODES.ERRORS.ERR_COORDINATOR_LOAD_IN_PROGRESS,
+        ];
+
+        this.consumerOptions = {
+            ...options,
+            consumeTimeout: options?.consumeTimeout ?? 1000,
+            consumeLoopTimeoutDelay: options?.consumeLoopTimeoutDelay ?? 100,
+            optimizedRebalance: options?.optimizedRebalance ?? false,
+            commitRetryDelay: options?.commitRetryDelay ?? 1000,
+            automaticConsume: options?.automaticConsume ?? true,
+            maxConsumerCommitRetries: options?.maxConsumerCommitRetries ?? 10,
+        };
+    }
+
+    /**
+     * Returns true if the consumer is connected
+     */
+    public isConnected() {
+        return this.consumer?.isConnected() ? true : false;
+    }
+
+    /**
+     * Returns the offset of the latest consumsed message
+     */
+    public getLatestMessageOffset(partitionId: number): number | undefined {
+        return this.latestOffsets.get(partitionId);
+    }
+
+    protected connect() {
+        if (this.closed) {
+            return;
+        }
+
+        const zookeeperEndpoints = this.endpoints.zooKeeper;
+        if (zookeeperEndpoints && zookeeperEndpoints.length > 0 && this.consumerOptions.zooKeeperClientConstructor) {
+            const zooKeeperEndpoint = zookeeperEndpoints[Math.floor(Math.random() % zookeeperEndpoints.length)];
+            this.zooKeeperClient = new this.consumerOptions.zooKeeperClientConstructor(zooKeeperEndpoint);
+        }
+
+        // eslint-disable-next-line prefer-const
+        let consumer: kafkaTypes.KafkaConsumer;
```

Comment:
If we don't reassign this later, why not make it const instead of disabling the eslint rule?

Reply:
Because it needs to be declared ahead of time for use by the "rebalance_cb" property in the kafka consumer constructor options.